File: unit_tests/unittest.rb
#!/bin/ruby
#
# unittest.rb
#
# Copyright 2025 DigiCert Project Authors. All Rights Reserved.
# 
# DigiCert® TrustCore and TrustEdge are licensed under a dual-license model:
# - **Open Source License**: GNU AGPL v3. See: https://github.com/digicert/trustcore-test/blob/main/LICENSE
# - **Commercial License**: Available under DigiCert’s Master Services Agreement. See: https://github.com/digicert/trustcore-test/blob/main/LICENSE_COMMERCIAL.txt  
#   or https://www.digicert.com/master-services-agreement/
# 
# For commercial licensing, contact DigiCert at sales@digicert.com.*
#
# This script will build a main.c file to run all the unit tests in a directory

require 'pathname'


# Lets us refine down the number of test sources we want to include by giving 
# cproto the concept of include dirs.  Without this the globbing is more fuzzy 
# and just includes everything
if ARGV.length == 4
  # Dir of this file
  m_unit_test_dir = File.expand_path(File.dirname(__FILE__))
  m_mss_dir = File.expand_path(m_unit_test_dir + '/..')
  # Start building the include-path for cproto
  m_include_dirs = Array.new
  m_include_dirs.push('-I ' + m_mss_dir + '/src')
  m_include_dirs.push('-I ' + m_mss_dir + '/src/common')
  m_include_dirs.push('-I ' + m_mss_dir + '/src/asn1')
  m_include_dirs.push('-I ' + m_mss_dir + '/src/platform')
  m_include_dirs.push('-I ' + m_mss_dir + '/src/crypto')
  m_include_dirs.push('-I ' + m_mss_dir + '/src/crypto/pqc')
  m_include_dirs.push('-I ' + m_mss_dir + '/src/crypto_interface')
  m_include_dirs.push('-I ' + m_mss_dir + '/src/data_protection')
  m_include_dirs.push('-I ' + m_mss_dir + '/thirdparty/bssl/include')
  m_include_dirs.push('-I ' + m_mss_dir + '/thirdparty/wolfssl-5.7.2/')

  $m_includes_str = ""
  m_include_dirs.each do |item|
    $m_includes_str = $m_includes_str + ' ' + item
  end
end

#########################
# This function collects all the function that match the prototype
# specified
def collect_functions(fileName, opts)
  is_windows = (ENV['OS'] == 'Windows_NT')
  result = Array.new
  funPrefix = File.basename(fileName, File.extname(fileName))
  puts "\n/*============#{fileName}===========*/\n\n"
  re = Regexp.new("int\\s*(" + funPrefix + "\\w*)\\s*\\(")
  # use cproto to collect the prototypes
  out_dir = "/dev/null"
  if is_windows
    out_dir = "nul"
  end

  # Continuation of the include dir logic mentioned above. Use our include dirs 
  # in the call to cproto
  if ARGV.length == 4
    if opts.nil?
      prototypes = `cproto -E cpp #{$m_includes_str} -O #{out_dir} #{fileName}`.split("\n")
    else
      prototypes = `cproto #{opts} -E cpp #{$m_includes_str} -O #{out_dir} #{fileName}`.split("\n")
    end
  else
    if opts.nil?
      prototypes = `cproto -E cpp -O #{out_dir} #{fileName}`.split("\n")
    else
      prototypes = `cproto #{opts} -E cpp -O #{out_dir} #{fileName}`.split("\n")
    end
  end

  prototypes.each do |prototype|
    if md = re.match(prototype)
      puts "/* found #{prototype} ====> function is #{md[1]} */"
      result.push(md[1]) # store the function name
    end
  end
  puts "/*==================================*/\n\n"
  return result
end # collect_functions


#########################


if ARGV.length == 0
  puts "Please specify a directory"
  return 2
end

## Get the host id from the args
hostid = ARGV[1]
opts = ARGV[2]

## Get the list of test sources if available
if ARGV.length == 4
  test_srcs_list = Array.new
  File.readlines(ARGV[3]).each do |line|
    # Do not include commented lines
    unless line.to_s.start_with?('#')
      test_srcs_list.push(line.to_s)
    end
  end
else
  test_srcs_list = nil
end

# Get just the basename of the test sources
unless test_srcs_list.nil?
  test_srcs_list.map! { |src| './' + File.basename(src).delete("\n") }
end

# header -- add more
puts "/* main.c \n*\n* test driver generated on #{Time.now} \n*/\n\n"
unit_test_path = Pathname.new("../../../unit_tests")
src_path = Pathname.new(ARGV[0])
rel_path = unit_test_path.relative_path_from(src_path)

puts "#if defined(__RTOS_WIN32__)"
puts "#include \<winsock2.h\>\n"
puts "#pragma comment(lib,\"ws2_32.lib\")\n"
puts "#endif"

puts "#ifdef __UNITTEST_REMOTE_SUPPORT__"
puts "#include \<sys/types.h\>\n"
puts "#include \<sys/socket.h\>\n"
puts "#endif"

puts "#include \"#{rel_path}/unittest.h\"\n\n"

fileFuns = Hash.new

# C files
Dir[ARGV[0]+"/*_test.c"].each do |path|
  # read the file and look for functions like "int fileName_test_xyz()"
  if test_srcs_list.nil?
    fileFuns[File.basename(path)]= collect_functions(path, opts)
  else
    # If we have populated the test_srcs list then its a subset test so only
    # collect functions if its in test_srcs
    if test_srcs_list.include? path
      fileFuns[File.basename(path)]= collect_functions(path, opts)
    end
  end
end

# C++ files
Dir[ARGV[0]+"/*_test.cpp"].each do |path|
  # read the file and look for functions like "int fileName_test_xyz()"
  fileFuns[File.basename(path)]= collect_functions(path, opts)
end

# prototype
fileFuns.each_key do |fileName|
  puts "\n\n/* functions in file #{fileName} */"
  puts fileFuns[fileName].map { |f| "int #{f}();" }.join("\n")
  puts "\n"
end

# TestDescriptors
puts <<EOTD
TestDescriptor gTestDescs[] = {
EOTD
fileFuns.each_key do |fileName|
  fileFuns[fileName].each do |function|
    puts "\n\tTEST_DESC(\"#{fileName}\", #{function}),"
  end
end

# main

puts <<EOH
};

#define SECS_TO_WAIT_FOR_TARGET 60
int main_host(int argc, char* argv[])
{
  int retVal = 0;
  int pass = 0;
  int connfd = 0;
  int totalTests = sizeof(gTestDescs)/sizeof(gTestDescs[0]) ;

  retVal = CONNECT_TEST_TARGET_H("#{hostid}", "test", argc, argv, SECS_TO_WAIT_FOR_TARGET, &connfd);
  if (retVal != 0)
      return retVal;

  if ( argc > 1)
  {
     totalTests = 0;
     retVal = RUN_TEST_BY_NAMES_H( connfd, (const char**)argv+1, argc-1, gTestDescs,
              sizeof(gTestDescs)/sizeof(gTestDescs[0]), &pass, &totalTests);
  } else
  {
EOH

fileFuns.each_key do |fileName|
  fileFuns[fileName].each do |function|
    puts "\n\tretVal += RUN_TEST_H( connfd, \"#{fileName}\", #{function}, &pass);"
  end
end
puts <<EOT
  }

  STOP_TEST_TARGET_H(&connfd, 1);

  report(totalTests, pass);
  return retVal;
}

#ifdef __UNITTEST_REMOTE_RUNTARGET__
int main_target(int argc, char* argv[])
{
  int retVal = 0;
  int connfd = 0;
  struct sockaddr hostip = { 0 };

  retVal = INITCOMM_TGT(&hostip, &connfd);
  if (retVal != 0)
      return retVal;

  retVal = PROC_RUNTEST_CMDS_TGT(&hostip, connfd, gTestDescs, sizeof(gTestDescs)/sizeof(gTestDescs[0]) );

  STOPCOMM_TGT(&connfd, retVal);

  return retVal;
}
#endif

int main(int argc, char* argv[])
{

#ifdef __UNITTEST_REMOTE_RUNTARGET__
  return main_target(argc,argv); /* run the Target. */
#else
  return main_host(argc,argv); /* run the Host. */
#endif

}

EOT

File: unit_tests/unittest.c
/*
 *  unittest.c
 * 
 *   unit test support
 *
 * Copyright 2025 DigiCert Project Authors. All Rights Reserved.
 * 
 * DigiCert® TrustCore and TrustEdge are licensed under a dual-license model:
 * - **Open Source License**: GNU AGPL v3. See: https://github.com/digicert/trustcore-test/blob/main/LICENSE
 * - **Commercial License**: Available under DigiCert’s Master Services Agreement. See: https://github.com/digicert/trustcore-test/blob/main/LICENSE_COMMERCIAL.txt  
 *   or https://www.digicert.com/master-services-agreement/
 * 
 * For commercial licensing, contact DigiCert at sales@digicert.com.*
 *
 */

#include "unittest.h"
#include "unittest_remote.h"
#include "../src/common/mtypes.h"
    
#if (defined(WIN32) && defined(_DEBUG))
#include <CrtDbg.h>
#endif

#include <assert.h>
#if defined( __RTOS_WIN32__) || defined( __RTOS_LINUX__) || \
    defined(__RTOS_CYGWIN__) || defined( __RTOS_SOLARIS__) || \
    defined(__RTOS_VXWORKS__) || defined( __RTOS_OSX__) || \
    defined(__RTOS_OPENBSD__) || defined( __RTOS_FREEBSD__) || \
    defined(__RTOS_IRIX__) || defined( __RTOS_DUMMY__ ) || \
    defined(__RTOS_SYMBIAN32__)  || defined(__RTOS_WINCE__) || \
    defined(__RTOS_QNX__)
#include <stdio.h>
#define PRINTF      printf
#else
/* OSes with no printf go here and need to define equivalent functionality*/
/* need to support only %s and %d format strings with no extensions */
#endif

/************************************
// Test execution debug-io support.
*************************************/
#define UT_DBG_COMM_HIGH_V   1 /* 0 = OFF 1 = ON */


/*-------------------------------------------------------------------------*/
#ifdef __UNITTEST_REMOTE_SUPPORT__


#define UT_DEBUG_PRINT(t,b,c) ut_debug_print(__FILE__, __LINE__, t, b, c)
static void ut_debug_print(const char* pfile, int iline, int printthis, char *str1, char *str2)
{
    char *pNull = "null";
    if (printthis == 0)
        return;
    if (!str1) str1 = pNull;
    if (!str2) str2 = pNull;
    printf("%s : %d : %s%s\n",pfile,iline,str1,str2);
    
}

#define MAX_DEBUGMSG_SIZE (4*1024)
char debugmsg_spot[MAX_DEBUGMSG_SIZE];
#endif /* __UNITTEST_REMOTE_SUPPORT__ */

int remote_target_socket = 0;	/* Needed by tests to determine if the peer IP address should be used of if localhost is acceptable */

int unittest_intEQ(const char* file, int line, const char* test, 
                int hint, int result, int expected)
{
#ifndef __UNITTEST_REMOTE_SUPPORT__

    if ( result != expected)
    {
        PRINTF("%s:%d:Test %s (%d) (%08x) result = %d, expected = %d\n", 
            file, line, test, hint, hint, result, expected);
        return 1;
    }
    return 0;
#else
    if ( result != expected)
    {
        sprintf(debugmsg_spot,"%s:%d:Test %s (%d) (%08x) result = %d, expected = %d\n",
            file, line, test, hint, hint, result, expected);
        unittest_write(debugmsg_spot);
        return 1;
    }
    return 0;
#endif
}


/*-------------------------------------------------------------------------*/

int unittest_intNE(const char* file, int line, const char* test, 
                int hint, int result, int unExpected)
{
#ifndef __UNITTEST_REMOTE_SUPPORT__
    if ( result == unExpected)
    {
        PRINTF("%s:%d:Test %s (%d)(%08x) result = %d, expected != %d\n",
            file, line, test, hint, hint, result, unExpected);
        return 1;
    }
    return 0;
#else
    if ( result == unExpected)
    {
        sprintf(debugmsg_spot,"%s:%d:Test %s (%d)(%08x) result = %d, expected != %d\n",
            file, line, test, hint, hint, result, unExpected);
        unittest_write(debugmsg_spot);
        return 1;
    }
    return 0;
#endif

}


/*-------------------------------------------------------------------------*/

int unittest_intGE( const char* file, int line, const char* testExpr, 
            int hint, int result, int expected)
{
#ifndef __UNITTEST_REMOTE_SUPPORT__
   if ( result < expected)
    {
        PRINTF("%s:%d:Test %s (%d)(%08x) result = %d, expected >= %d\n", 
            file, line, testExpr, hint, hint, result, expected);
        return 1;
    }
    return 0;
#else
    if ( result < expected)
     {
        sprintf(debugmsg_spot,"%s:%d:Test %s (%d)(%08x) result = %d, expected >= %d\n",
             file, line, testExpr, hint, hint, result, expected);
        unittest_write(debugmsg_spot);
         return 1;
     }
     return 0;
#endif
}


/*-------------------------------------------------------------------------*/

void unittest_write(const char* msg)
{
#ifndef __UNITTEST_REMOTE_SUPPORT__
    PRINTF("%s",msg);
#else
    if (remote_target_socket == 0)
    {
    	/* Host side. */
        PRINTF("%s",msg);
    }
    else
    {
    	/* Target side. */
        ut_write_output_cmd_tgt(remote_target_socket, msg);
    }
#endif
}


/*-------------------------------------------------------------------------*/

char* addhexChar( char* out, const unsigned char c)
{
    int n = (c >> 4) & 0x0F;
    *out++ = (char) ( (n < 0xA) ? '0' + n : 'A' + n - 0x0A);
    n = c & 0xF;
    *out++ =  (char) ( (n < 0xA) ? '0' + n : 'A' + n - 0x0A);
    return out;
}

/*-------------------------------------------------------------------------*/

void unittest_write_buffer( const unsigned char* buff, int buffLen)
{
    char line[82];
    line[80] = '\n';
    line[81] = 0;
    /* 6 groups of 8 hex + space = 54 +
       1 group of 24 chars = 24 -> 78 chars */

    while (buffLen)
    {
        char* out = line;
        int i, maxChars = ( buffLen > 24) ? 24 : buffLen;

        out = line;
        for (i = 0; i < maxChars; ++i)
        {
            out = addhexChar(out, buff[i]);
            if (i % 4 == 3)
            {
                *out++ = ' ';
            }
        }
        while (out != line +54)
        {
            *out++ = ' ';
        }

        for (i = 0; i < maxChars; ++i)
        {
            char c = buff[i];
            *out++ = (c >= 0x20 && c <= 0x7E) ? c : '.';
        }
        while (out != line + 80)
        {
            *out++ = ' ';
        }

        unittest_write(line);

        buff += maxChars;
        buffLen -= maxChars;
        
    }
}


/*-------------------------------------------------------------------------*/

int run_test( const char* testFunName, const char* fileName,
             TestFun testFun, int* pass)
{
    int retVal;

    
#if (defined(WIN32) && defined(_DEBUG))
    _CrtMemState memState;
    /* Get current flag */
    int tmpFlag = _CrtSetDbgFlag( _CRTDBG_REPORT_FLAG );
    /* Turn on extensive checks */
    tmpFlag |= (_CRTDBG_CHECK_ALWAYS_DF | _CRTDBG_LEAK_CHECK_DF);

    /* Set flag to the new value */
    _CrtSetDbgFlag( tmpFlag );
   _CrtSetReportMode( _CRT_WARN, _CRTDBG_MODE_DEBUG );
   _CrtSetReportMode( _CRT_ERROR, _CRTDBG_MODE_DEBUG );
   _CrtSetReportMode( _CRT_ASSERT, _CRTDBG_MODE_DEBUG );

   _CrtMemCheckpoint( &memState);

#endif
    
    PRINTF("Running test %s in file %s\n", testFunName, fileName);
#ifndef __RTOS_QNX__
    fflush(stdout);
#endif
    retVal = testFun();
#if (defined(WIN32) && defined(_DEBUG))
    if (!_CrtCheckMemory()) ++retVal;
    _CrtMemDumpAllObjectsSince(&memState);
#endif
    if ( 0==retVal)
    {
        PRINTF("%s: Pass\n\n", testFunName);
        ++(*pass);
    }
    else
    {
        PRINTF("%s: Fail: %d error(s)\n\n", testFunName, retVal);
    }
#ifndef __RTOS_QNX__
    fflush(stdout);
#endif

    return retVal;
}


/*-------------------------------------------------------------------------*/

void report( int total, int pass)
{
    PRINTF(  "\nPass:        %d", pass);
    PRINTF(  "\nFail:        %d", total - pass);
    PRINTF(  "\nTotal tests: %d\n", total);
}


/*-------------------------------------------------------------------------*/

int run_test_by_name( const char* name, TestDescriptor* tests, 
                      int num_tests, int* pass, int* total_tests)
{
    int retVal = 0;
    int i, j, name_len;
    const char* p;

    name_len = 0;
    p = name;

    while (*p) { ++name_len; ++p; }
    if (name_len)
    {
        for (i = 0; i < num_tests; ++i)
        {
            const char* testName;

            testName = tests[i].testName;

            for (j = 0; j < name_len; ++j)
            {
                if (testName[j] != name[j])
                {
                    break;
                }
            }
            if ( name_len == j) /* match */
            {
                (*total_tests)++;
                retVal += run_test(tests[i].testName, tests[i].fileName,
                                   tests[i].testFun, pass);
            }
        }
    }
    return retVal;
}


/*-------------------------------------------------------------------------*/

int list_test_names( TestDescriptor* tests, int num_tests)
{
    int i;

    PRINTF("Available tests:\n\n");
    for (i = 0; i < num_tests; ++i)
    {
        PRINTF("%s",tests[i].testName);
        PRINTF("\n");
    }

    return 0;
}


/*-------------------------------------------------------------------------*/

int run_test_by_names( const char* names[], int num_names, 
                       TestDescriptor* tests, int num_tests, 
                       int* pass, int* total_tests)
{
    int retVal = 0;
    int i;
    
    if (1 == num_names)
    {
        const char* option = names[0];

        if ('-' == option[0]  && 'h' == option[1]  && 0 == option[2] )
        {
            list_test_names(tests, num_tests);
            return 0;
        } 
    }

    for (i = 0; i < num_names; ++i)
    {
        retVal += run_test_by_name( names[i], tests, num_tests, pass, 
                                    total_tests);
    }
    return retVal;
}


#if 0
/* Disable until testmonkey infrastructure could be fixed to handle option
  flags
*/
#include "../src/common/moptions.h"
#include "../src/crypto/hw_accel.h"
#if defined(__ENABLE_HARDWARE_ACCEL_ASYNC_CRYPTO__)
static int utest_hwAccelCtx  = 0;
static int hwAccelRefCount = 0;

int
UNITTEST_getHwAccelChannel(void)
{
    if (0 == utest_hwAccelCtx) {
        HARDWARE_ACCEL_OPEN_CHANNEL(9, &utest_hwAccelCtx);
    }
exit:
    assert(utest_hwAccelCtx != 0);
    return utest_hwAccelCtx;
}

void
UNITTEST_releaseHwAccelChannel(int hwAccelCtx)
{
}
#endif
#endif

#ifdef __UNITTEST_REMOTE_SUPPORT__

/*********************************
 Real remote support.
************************************/
int connect_test_target_h(const char *tgtname, const char *execname,
						  int argc, char* argv[], int timeoutsecs, int *pconnfd)
{
	int rc = 0;
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"connect_test_target_h: target name=",tgtname);

    *pconnfd = 0;

    remote_target_socket = 0; /* Host side doesn't use one. */

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"connect_test_target_h: "," Load test target.");
	rc = ut_load_test_target_h(tgtname, execname, argc, argv, timeoutsecs);  /* May exec on localhost, may be a nop. */

	if (rc == 0)
	{
		UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"connect_test_target_h: "," Connect test target.");
		rc = ut_connect_test_target_h(tgtname, execname, argc, argv, timeoutsecs, pconnfd); /* Connect via socket. */
	}

	if (rc == 0)
	{
		UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"connect_test_target_h: "," Do Startup cmd.");
		rc = ut_startup_cmd_h(*pconnfd);
	}

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"connect_test_target_h"," done.");
	return rc;
}

int stop_test_target_h(int *pconnfd, int timeoutsecs)
{
	int rc = -1;
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"stop_test_target_h"," ...");
    if (*pconnfd != 0)
    {
    	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"stop_test_target_h"," Sending shutdown cmd.");
    	rc = ut_shutdown_cmd_h(*pconnfd, timeoutsecs); /* Send Shutdown cmd and wait for ACK. */

    	rc = ut_disconnect_test_target_h(pconnfd); /* Closes socket. */
    }

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"stop_test_target_h"," done.");
	return rc;
}

int run_test_h( int connfd, const char* testFunName, const char* fileName,
             TestFun testFun, int* pass)
{
	int rc = -1;
	int retVal = 0;

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"run_test_h"," ...");
    if (connfd != 0)
    {
	    PRINTF("Remotely Running test %s in file %s\n", testFunName, fileName);
    	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"run_test_h"," Sending run_test cmd.");
    	rc = ut_runtest_cmd_h(connfd, testFunName, fileName, &retVal); /* Send RUNTEST cmd, and wait until _FINISHED received.*/
    }

    if ((0==retVal) && (0==rc))
    {
        PRINTF("%s: Pass\n\n", testFunName);
        ++(*pass);
    }
    else
    {
    	if (rc != 0)
    	{
    		retVal = rc; /* This is to catch comm errors.*/
    	}
        PRINTF("%s: Fail: %d error(s)\n\n", testFunName, retVal);
    }

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"run_test_h"," done.");
  	return retVal;

}

int run_test_by_name_h( int connfd, const char* name, TestDescriptor* tests,
                      int num_tests, int* pass)
{
	int index = -1;
	int retVal = 0;
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"run_test_by_name_h"," ...");

	index = ut_testindexfromstr(tests, num_tests, name);
	if (index >= 0)
	{
		retVal += run_test_h(connfd, tests[index].testName, tests[index].fileName,
	                         tests[index].testFun, pass);
	}
	else
	{
        PRINTF("%s: Fail: Test not found.\n", name);
	}

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"run_test_by_name_h"," done.");
  	return 0;
}

int run_test_by_names_h( int connfd, const char* names[], int num_names,
                       TestDescriptor* tests,
                       int num_tests, int* pass)
{
    int retVal = 0;
    int i;

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"run_test_by_names_h"," ...");
    if (1 == num_names)
    {
        const char* option = names[0];

        if ('-' == option[0]  && 'h' == option[1]  && 0 == option[2] )
        {
            list_test_names(tests, num_tests);
            return 0;
        }
    }

    for (i = 0; i < num_names; ++i)
    {
        retVal += run_test_by_name_h(connfd, names[i], tests, num_tests, pass);
    }

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"run_test_by_names_h"," done.");
    return retVal;

}


int initcomm_tgt( struct sockaddr *phostip, int *pconnfd)
{
	int rc;

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"initcomm_tgt"," ...");
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"initcomm_tgt"," Accept incoming connect.");

	remote_target_socket = 0; /* Target side I/O needs this. But default to Zero... */

    rc = ut_initcomm_tgt(phostip, pconnfd);  /* Listen for incoming connections, and accept, returning connected socket. */
    if (rc == 0)
    {
    	remote_target_socket = *pconnfd; /* Target side I/O needs this.*/

    	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"initcomm_tgt"," Wait for startup cmd.");
    	rc = ut_waitfor_startup_cmd_tgt(*pconnfd);
    }

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"initcomm_tgt"," done.");
	return rc;
}

int proc_runtest_cmds_tgt( struct sockaddr *phostip, int connfd, TestDescriptor* tests, int num_tests)
{
	int rc;
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"proc_runtest_cmds_tgt"," ...");
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"proc_runtest_cmds_tgt"," Wait for runtest cmds.");
	rc = ut_waitfor_runtest_cmd_tgt(connfd, tests, num_tests);
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"proc_runtest_cmds_tgt"," done.");
    return rc;
}

int stopcomm_tgt( int *pconnfd, int statusin)
{
	int rc;
	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"stopcomm_tgt"," ...");

	rc = ut_stopcomm_tgt(pconnfd, statusin);

    remote_target_socket = 0; /* Target side I/O needs this. But default to Zero... */

	UT_DEBUG_PRINT(UT_DBG_COMM_HIGH_V,"stopcomm_tgt"," done.");
	return rc;
}

#endif  /*  __UNITTEST_REMOTE_SUPPORT__ */

#if defined(__ENABLE_DIGICERT_UNITTEST__)

#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>

char *gpOutFile[256];
int gOutFD;
int stdoutFD;

void redirectOutput(char *pFile)
{
    fflush(NULL);
    memcpy(gpOutFile, pFile, strlen(pFile));
    gpOutFile[strlen(pFile)] = '\0';
    stdoutFD = dup(1);
    gOutFD = open(pFile, O_CREAT|O_WRONLY|O_TRUNC, 0666);
    dup2(gOutFD, 1);
}

void restoreOutput()
{
    fflush(NULL);
    dup2(stdoutFD, 1);
    close(stdoutFD);
}

#endif /* __ENABLE_DIGICERT_UNITTEST__ */